File: CodingTest-2026/tower-of-hanoi/Solution2.java
// Solution2: 이동 번호 기반 시뮬레이션 방식

import java.io.*;

class Main {

    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        long K = Long.parseLong(br.readLine());

        // 처음에는 모든 원반이 첫 번째 막대에 있다.
        // 1 + 2 + ... + 20 = 210
        int[] rodSum = {210, 0, 0};

        // position[disk] = 현재 원반이 위치한 막대 번호
        // 초기값 0이므로 모든 원반은 첫 번째 막대에서 시작한다.
        int[] position = new int[21];

        for (long move = 1; move <= K; move++) {

            // 이동 번호의 trailing zero 개수 + 1이
            // 현재 움직이는 원반의 크기이다.
            int disk = Long.numberOfTrailingZeros(move) + 1;

            // 20개의 원반에서는
            // 홀수 원반은 +1 방향,
            // 짝수 원반은 -1 방향으로 순환한다.
            int direction;

            if (disk % 2 == 1) {
                direction = 1;
            } else {
                direction = -1;
            }

            int current = position[disk];
            int next = (current + direction + 3) % 3;

            // 기존 막대에서 원반을 제거한다.
            rodSum[current] -= disk;

            // 새로운 막대에 원반을 추가한다.
            rodSum[next] += disk;

            // 원반의 현재 위치를 갱신한다.
            position[disk] = next;
        }

        System.out.println(
            rodSum[0] + " " + rodSum[1] + " " + rodSum[2]
        );
    }
}

File: CodingTest-2026/tower-of-hanoi/Solution2.py
### Solution2: 이동 번호 기반 시뮬레이션 방식 ###

K = int(input())

# 모든 원반의 초기 합
rod_sum = [210, 0, 0]

# position[disk] = 현재 원반의 막대 번호
position = [0] * 21

for move in range(1, K + 1):

    # move의 가장 낮은 1비트가
    # 현재 움직이는 원반을 결정한다.
    low_bit = move & -move
    disk = low_bit.bit_length()

    # 20개의 원반에서는
    # 홀수 원반은 +1 방향,
    # 짝수 원반은 -1 방향으로 움직인다.
    if disk % 2 == 1:
        direction = 1
    else:
        direction = -1

    current = position[disk]
    next_position = (current + direction) % 3

    # 기존 막대에서 제거
    rod_sum[current] -= disk

    # 새로운 막대에 추가
    rod_sum[next_position] += disk

    # 현재 위치 갱신
    position[disk] = next_position

print(*rod_sum)
